### ISSUE TITLE
movabs incorrect?

### ISSUE THREAD

**Issue by mlgiraud** [CensoredUsername/dynasm-rs#108]
Hi, im currently writing a jit compiler and need to load a 64 bit immediate value into a register. The only way to do this in one instruction afaik is by using the mov encoding `movabs rax, imm64`. This however currently emits code that should correspond to something like `movabs rax, [imm64]`, i.e. it tries to load from the immediate address. These encodings exist, but should be generated when writing `movabs rax, [imm64]` imho.

The language spec in the documentation says:
```
movabs al, imm64
movabs ax, imm64
movabs eax, imm64
movabs imm64, al
movabs imm64, ax
movabs imm64, eax
movabs imm64, rax
movabs rax, imm64
```

Which should probably be
```
movabs al, [imm64]
movabs ax, [imm64]
movabs eax, [imm64]
movabs [imm64], al
movabs [imm64], ax
movabs [imm64], eax
movabs [imm64], rax
movabs rax, [imm64]

movabs reg64, imm64 <---- This is missing
```

The spec in the documentation says that `mov reg64, imm64` can be used, but this results in an error (or the truncation of imm64 to a 32 bit value if you leave the type up to the compiler with `as _`).
I think this should be moved to `movabs reg64, imm64`.

**Comment by mlgiraud** [comment]
Yeah i see the reasoning and agree that it is probably nicer to use the `mov reg, QWORD imm` syntax. However, i would still suggest changing the movabs syntax to take the address via `[imm64]`. This was confusing for me since most (all?) instruction use the `[]` syntax when an address is involved, right? Also, other assemblers/disassemblers use the same syntax with the `[]`. This will inevitably lead to confusion imho.

**Comment by CensoredUsername** [comment]
I understand that this would be clearer. Unfortunately with how the x64 backend works right now this would entail a pretty big rewrite of the parser/compiler. It is built towards memory references only being of the regular kind, which have a specific instruction structure, while 64-bit displacement mov is actually encoded as a single register + 64-bit immediate operand.

**Comment by mlgiraud** [comment]
Yeah i suspected as much, maybe we can add a small disclaimer in the documentation for x86 in the movabs part?

**Comment by CensoredUsername** [comment]
I've added a disclaimer in the docs in 55e83aa as I needed to revisit this part for label support in the 64-bit immediate/displacement instructions.

**Comment by mlgiraud** [comment]
Perfect. I wouldn't have worded it in a different way. Closing this. Thanks for your work.
